### DISCUSSION TITLE
Disallow using promise without a done.fail

### DISCUSSION THREAD

**Discussion by thabalija** [infinum/js-linters#53] [General]
We should prevent commiting fdescribe/fit.

```js
{
  "overrides": [
    {
      "plugins": ["jasmine"],
      "rules": ["jasmine/no-promise-without-done-fail": 2]
    }
  ]
}
```

Jasmine plugin: [eslint-plugin-jasmine](https://github.com/tlvince/eslint-plugin-jasmine)

**Comment by fvoska** [comment]
I think this rule is ok if you are using .then() because this ensures that the test waits until the promise is resolved and does not move on to the next test before resolution, as this can cause all sorts of issues.

However, I think it is even better to use async/await when working with promises in tests. That way you don't even need `done`. Basically, I would like to minimize the usage of `done` and utilize async/await as much as possible.

Example before:

```ts
describe('A suite', function() {
  it('A spec', function(done) {
    asyncCall().then(onSuccess, done.fail);
  });
});
```

Example after:

```ts
describe('A suite', () => {
  it('A spec', async () => {
    const result = await asyncCall();

    expect(...)...
  });
});
```

If the promise is rejected, the test will automatically fail with an exception because you do not have a try-catch block. No need to re-throw explicitly using done.fail.

**Comment by fvoska** [comment]
When working with observables, you have 2 options. You can either convert it to a promise and await, or you can use `subscribe` in a similar manner as you would use `then` with promises. There are some cases where one is simpler than the other. If subscribing, you also have to add `done.fail` in the subscriber's error handler. I am not sure if this is caught by this rule (probably not)

**Comment by thabalija** [comment]
I have opened [PR](https://github.com/infinum/js-linters/pull/56) for using jasmine plugin recommended settings.
